### PR TITLE
React Upgrade

### LINKED ISSUE #2550: React Upgrade

### PR DESCRIPTION
**Pull request by qweliant**
## Issue(s) Resolved
#2550
Bumps React to next major version

## Test Plan
Remove node_modules folder and run `npm i`
Run all test locally
Start Storybook and navigate to be sure it still runs as expected
Run your local server and navigate around the site being sure nothing changes

Visit the SubmissionWorkflow editor and be sure stepping through the tabs works

Visit a pub and release it. Create a comment. Be sure the time listed has not been affected by the requirement of the `title` prop.

### Notes/Context/Gotchas
Note title for `TimeAgo` is [not required](https://github.com/nmn/react-timeago#title-optional) yet we still get a type err for it being present. See comment for screenshot below

React 17 does not introduce any changes to what we do. This means we should probably use the upgrade as intended and try going to r18.

### Supporting Docs
https://reactjs.org/blog/2020/08/10/react-v17-rc.html

### CODE REVIEW

File: server/Html.tsx [hunk 1]
```diff
@@ -101,7 +101,7 @@ const Html = (props: Props) => {
 				className={classNames(
 					props.bodyClassPrefix && `${props.bodyClassPrefix}-body-wrapper`,
 					getActiveSlugClassName(props.initialData, props.viewData),
-					getUserClassName(props.initialData),
+					{ ...getUserClassName(props.initialData) },
```
Comment (qweliant) [inline server/Html.tsx:104]: not sure why this wasn't flagged earlier but had to spread classnames bc return type from `getUserClassName` was of type String[]

File: client/components/GlobalControls/GlobalControlsButton.tsx [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import React from 'react';
 import classNames from 'classnames';
-import { Button, IButtonProps, AnchorButton } from '@blueprintjs/core';
+import { Button, ButtonProps, AnchorButton } from '@blueprintjs/core';
```
Comment (qweliant) [inline client/components/GlobalControls/GlobalControlsButton.tsx:3]: IButtonProps were deprecated and have updated them wear needed

File: client/components/GlobalControls/GlobalControlsButton.tsx [hunk 3]
```diff
@@ -55,10 +55,14 @@ const GlobalControlsButton = React.forwardRef((props: Props, ref: React.Ref<unkn
 			large: !isMobileButton,
 			className: classNames(className, renderProps.className),
 		};
+
+		// ? need to tell props that it will be buttonprops explicitly
```
Comment (qweliant) [inline client/components/GlobalControls/GlobalControlsButton.tsx:59]: I have this question mark here but feel there is a better way to express the two props. Also not sure why I need to update this as the only bp package that needed an upgrade is bp select

File: client/components/Menu/Menu.tsx [hunk 4]
```diff
@@ -73,8 +73,8 @@ export const Menu = React.forwardRef((props: MenuProps, ref) => {
 				as="ul"
 				style={{ zIndex: 20, ...menuStyle }}
 				className={classNames(Classes.MENU, Classes.ELEVATION_1, className)}
-				unstable_portal={menuConfig.usePortal}
 				{...menu}
+				modal={menuConfig.usePortal}
```
Comment (qweliant) [inline client/components/Menu/Menu.tsx:77]: It seems `unstable_portal` has been deprecated for declaring if things will render in modal

File: client/components/TimeAgo/TimeAgo.tsx [hunk 5]
```diff
@@ -8,7 +8,7 @@ type Props = {
 	date: number | string | Date;
 	useDateCutoffDays?: number;
 	className?: string;
-	now?: number | string | Date;
+	now?: number;
```
Comment (qweliant) [inline client/components/TimeAgo/TimeAgo.tsx:11]: now can only be a number type, and if you go up the call stack youll see a number is only ever passed into `now`

File: client/containers/DashboardSubmissionWorkflow/SubmissionWorkflowStepPicker.tsx [hunk 6]
```diff
@@ -51,7 +51,6 @@ const SubmissionWorkflowStepPicker = (props: Props) => {
 							reached && 'reached',
 							incomplete && 'incomplete',
 						)}
-						stopId={step}
```
Comment (qweliant) [inline client/containers/DashboardSubmissionWorkflow/SubmissionWorkflowStepPicker.tsx:54]: This prop no longer exist on the reakit `Tab` component. I have tested the SW stepper and it seems non breaking

File: utils/storybook/stories.test.js [hunk 7]
```diff
@@ -1,7 +1,7 @@
 /**
  * @jest-environment jsdom
  */
-import Adapter from 'enzyme-adapter-react-16';
+import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
```
Comment (qweliant) [inline utils/storybook/stories.test.js:4]: completely diff package from 16 needed to be installed

File: client/components/Popover/PopoverButton.tsx [hunk 8]
```diff
@@ -38,9 +38,9 @@ const PopoverButton = (props: Props) => {
 			<Popover
 				aria-label={ariaLabel}
 				className={className}
-				unstable_portal={true}
 				tabIndex={0}
 				{...popover}
+				modal={true}
```
Comment (qweliant) [inline client/components/Popover/PopoverButton.tsx:43]: with the change and with `unstable_portal`

image

File: client/containers/DashboardEdges/NewEdgeInput.tsx [hunk 9]
```diff
@@ -1,6 +1,7 @@
 import React, { useState, useEffect } from 'react';
 import { MenuItem, Position, Spinner } from '@blueprintjs/core';
-import { Suggest } from '@blueprintjs/select';
+import { Suggest2 } from '@blueprintjs/select';
+// import { MenuItem2 } from "@blueprintjs/popover2"
```
Comment (howebenhowe) [inline client/containers/DashboardEdges/NewEdgeInput.tsx:4]: What is `MenuItem2`?

File: client/containers/Pub/PubHeader/Download.tsx [hunk 10]
```diff
@@ -156,7 +156,7 @@ const Download = (props: Props) => {
 						rightElement={
 							<span>
 								{isLoading && selectedType?.format === type.format && (
-									<Spinner size={Spinner.SIZE_SMALL} />
+									<Spinner size={SpinnerSize.SMALL} />
```
Comment (howebenhowe) [inline client/containers/Pub/PubHeader/Download.tsx:159]: Had to read this three times to see how it had changed, lol.

File: utils/storybook/stories.test.js (hunk 7 above)
Comment (howebenhowe) [reply utils/storybook/stories.test.js:4]: Glad for this adapter! Curious how we might chart into R18?
First [google result](https://dev.to/wojtekmaj/enzyme-is-dead-now-what-ekl) I'm chewing on.

**Review by howebenhowe — APPROVED** [review]
A couple consistency comments, but the code changes look great. Nice clean-up along the way, too!

There are a couple places where you move passing `modal={true}` prop to below where the results of `const popover = usePopoverState({...});` is spread (`<SomeComponent {...popover} modal={true} />`) and though I have _still_ been unable to `npm i` and test it myself, I think the modal can be configured through `usePopoverState` directly, rather than overwrite whatever comes out of the spread operator.

Then the bigger question of how to test with R18, but that's a question for another ticket altogether.

Got it building!

Storybook looks great, but complained when getting started with:
```WARN   Failed to load preset: {"type":"presets"} on level 1
ERR! TypeError [ERR_INVALID_ARG_TYPE]: The "id" argument must be of type string. Received an instance of Object
ERR!     at new NodeError (node:internal/errors:387:5)
```

clean run through submission workflow!

**Pull request closed without merging**